keras_cv/losses/giou_loss.py

```diff
@@ -0,0 +1,151 @@
+
+    References:
+        - [GIoU paper](https://arxiv.org/pdf/1902.09630)
+        - [TFAddons Implementation](https://www.tensorflow.org/addons/api_docs/python/tfa/losses/GIoULoss)
+    Sample Usage:
+    ```python
+    y_true = tf.random.uniform((5, 10, 5), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    y_pred = tf.random.uniform((5, 10, 4), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    loss = GIoULoss(bounding_box_format = "xyWH")
+    loss(y_true, y_pred).numpy()
+    ```
+    Usage with the `compile()` API:
+    ```python
+    model.compile(optimizer='adam', loss=keras_cv.losses.GIoULoss())
+    ```
+    """
+
+    def __init__(self, bounding_box_format, axis=None, **kwargs):
+        super().__init__(**kwargs)
+        self.bounding_box_format = bounding_box_format
+        self.axis = axis
+
+    def _compute_giou(self, boxes1, boxes2):
+        boxes1_rank = len(boxes1.shape)
+        boxes2_rank = len(boxes2.shape)
+
+        if (
+            boxes1_rank != boxes2_rank
+            or boxes1_rank not in [2, 3]
+            or boxes2_rank not in [2, 3]
+        ):
+            raise ValueError(
+                "`GIoULoss` expects both boxes to be batched, or both "
+                f"boxes to be unbatched.  Received `len(boxes1.shape)`={boxes1_rank}, "
+                f"`len(boxes2.shape)`={boxes2_rank}.  Expected either `len(boxes1.shape)`=2 AND "
+                "`len(boxes2.shape)`=2, or `len(boxes1.shape)`=3 AND `len(boxes2.shape)`=3."
+            )
+
+        if self.bounding_box_format.startswith("rel"):
+            target = "rel_yxyx"
+        else:
+            target = "yxyx"
+
+        boxes1 = bounding_box.convert_format(
+            boxes1, source=self.bounding_box_format, target=target
+        )
+
+        boxes2 = bounding_box.convert_format(
+            boxes2, source=self.bounding_box_format, target=target
+        )
+
+        def compute_giou_for_batch(boxes):
+            boxes1, boxes2 = boxes
+            zero = tf.convert_to_tensor(0.0, boxes1.dtype)
+            boxes1_ymin, boxes1_xmin, boxes1_ymax, boxes1_xmax = tf.unstack(
+                boxes1[..., :4], 4, axis=-1
+            )
+            boxes2_ymin, boxes2_xmin, boxes2_ymax, boxes2_xmax = tf.unstack(
+                boxes2[..., :4], 4, axis=-1
+            )
+            boxes1_width = tf.maximum(zero, boxes1_xmax - boxes1_xmin)
+            boxes1_height = tf.maximum(zero, boxes1_ymax - boxes1_ymin)
+            boxes2_width = tf.maximum(zero, boxes2_xmax - boxes2_xmin)
+            boxes2_height = tf.maximum(zero, boxes2_ymax - boxes2_ymin)
+            boxes1_area = boxes1_width * boxes1_height
+            boxes2_area = boxes2_width * boxes2_height
+            intersect_ymin = tf.maximum(boxes1_ymin, boxes2_ymin)
+            intersect_xmin = tf.maximum(boxes1_xmin, boxes2_xmin)
+            intersect_ymax = tf.minimum(boxes1_ymax, boxes2_ymax)
+            intersect_xmax = tf.minimum(boxes1_xmax, boxes2_xmax)
```

Comment:
reuse box.compute_iou if possible

Reply:
Discussed this with @LukeWood earlier. There's no way to reuse `compute_iou` without either making changes to the API or changing the return type (both not ideal alternatives for an exposed function)